app/portainer/environments/environment.service/index.ts

```diff
@@ -0,0 +1,295 @@
+import { AxiosError } from 'axios';
+
+import axios from '@/portainer/services/axios';
+import PortainerError from '@/portainer/error';
+
+import {
+  Environment,
+  EnvironmentGroupId,
+  EnvironmentId,
+  EnvironmentType,
+  EnvironmentSettings,
+  TagId,
+  TeamId,
+  UserId,
+} from '../types';
+
+import { arrayToJson, buildUrl } from './utils';
+
+interface EndpointsQuery {
+  search?: string;
+  types?: EnvironmentType[];
+  tagIds?: TagId[];
+  endpointIds?: EnvironmentId[];
+  tagsPartialMatch?: boolean;
+  groupId?: EnvironmentGroupId;
+}
+
+export async function getEndpoints(
+  start: number,
+  limit: number,
+  { types, tagIds, endpointIds, ...query }: EndpointsQuery = {}
+) {
+  if (tagIds && tagIds.length === 0) {
+    return { totalCount: 0, value: <Environment[]>[] };
+  }
+
+  const url = buildUrl();
+
+  const params: Record<string, unknown> = { start, limit, ...query };
+
+  if (types) {
+    params.types = arrayToJson(types);
+  }
+
+  if (tagIds) {
+    params.tagIds = arrayToJson(tagIds);
+  }
+
+  if (endpointIds) {
+    params.endpointIds = arrayToJson(endpointIds);
+  }
+
+  try {
+    const response = await axios.get<Environment[]>(url, { params });
+
+    const totalCount = response.headers['X-Total-Count'];
+
+    return { totalCount: parseInt(totalCount, 10), value: response.data };
+  } catch (e) {
+    const axiosError = e as AxiosError;
+
+    if (!axiosError.isAxiosError) {
```

Comment:
I'm not a big fan of this particular check being repeated multiple times, would it make sense to implement a helper function, maybe in the `axios.ts` file something like `parseAxiosError` and here do something shorter like
`catch(e) {
	throw parseAxiosError(e);
}`